onnxruntime/core/session/inference_session.cc

```diff
@@ -869,19 +865,31 @@ common::Status InferenceSession::TransformGraph(onnxruntime::Graph& graph,
                                                 const onnxruntime::GraphTransformerManager& graph_transformer_mgr,
                                                 const ExecutionProviders& providers,
                                                 KernelRegistryManager& kernel_registry_manager,
-                                                const InsertCastTransformer& insert_cast_transformer,
                                                 SessionState& session_state,
                                                 bool saving_model_in_ort_format) {
   // The transformer order:
   // 1. run level 1 optimizations. these only use ONNX operators.
-  // 2. partition nodes based on EP capabilities. EPs may fuse nodes during this process.
-  // 3. run all optimizations. level 2 and 3 optimizations use contrib ops.
-  // 4. insert cast nodes
-  // 5. insert copy nodes
+  // 2. ensure potential QDQ node units have unique DQ nodes (required transformer).
+  // 3. partition nodes based on EP capabilities. EPs may fuse nodes during this process.
+  // 4. run level 2+ optimizations. level 2 and 3 optimizations use contrib ops.
+  // 5. insert cast nodes (required transformer).
+  // 6. insert copy nodes (required transformer).
+
+  auto apply_transformer_once = [](const GraphTransformer& transformer, const logging::Logger& logger,
+                                   Graph& graph) {
+    bool modified = false;
+    return transformer.Apply(graph, modified, logger);
+  };
 
   // first apply execution provider independent level 1 graph optimizations.
-  ORT_RETURN_IF_ERROR_SESSIONID_(
-      graph_transformer_mgr.ApplyTransformers(graph, TransformerLevel::Level1, *session_logger_));
+  ORT_RETURN_IF_ERROR_SESSIONID_(graph_transformer_mgr.ApplyTransformers(graph, TransformerLevel::Level1, *session_logger_));
+
+  // ensure potential QDQ node units have unique DQ nodes
+  {
+    // TODO should it only run if session option session.disable_quant_qdq == 0?
```

Comment:
should we only run EnsureUniqueDQForNodeUnit if session option session.disable_quant_qdq == 0?

Reply:
I think so. If we're disabling QDQ I assume we want the fewest nodes possible and that we're not converting QDQ node units to quantized ops. 

Not sure what the use-case is for disabling though. @yufenglee can maybe provide more details on when that flag would be set.